Shopify/shopify-app-js
# Add prisma adaptor

## WHY are these changes introduced?
We want to make it easy to store sessions in prisma

## WHAT is this pull request doing?
Implements a Session Storage adaptor. Some notes about the approach:

### Schema.prisma
The partner must implement `schema.prisma` themself (See readme).  I think this is best because:

1. Developers will be used to looking at `schema.prisma` to understand the database
2. I did not find an idiomatic way to schema stitch.  E.g: Our adaptor provides part of the schema, but the rest is up to the partner
3. I can't imagine we will change this schema often, and if we can do we can provide very simple instructions in release notes.  E.g: 1) Update the schema to this 2) Run `npx prisma migrate dev`

### Type strategy
The adaptor receives the prisma client as an argument.  In the partners codebase this prisma client will be strongly typed. E.g, it will understand that that the prisma client satisfies the requirements of `PrismaSessionStorage`:

```
prisma.session.upsert(...)
prisma.session.delete(...)
prisma.session.deleteMany(...)
prisma.session.findUnique(...)
prisma.session.findMany(...)
```

But as soon as the partner adds a new table to the database the prisma client will also support other methods, e.g:

```
prisma.widgets.upsert(...)
```

This means that the prisma client you pass to the Session adaptor should satisfy at-least the session table requirements, but we don't care about anything else.  The way I've solved this is:

1. Create the minimum viable schema in the repo
2. Generate the prisma client.
3. Extend the types generated by the prisma client with unknown keys, e.g: `PrismaClient | (PrismaClient & {[key: string]: unknown})`.

I'd love input here.  Some concerns I have:

1. You need to run `npx prisma migrate dev` to get the types from the Prisma client.  I don't see how that is run before we build the package.  If it isn't I'm not sure what the types will be for an external consumer.
2. PrismaClient | (PrismaClient & {[key: string]: unknown}) is a little ugly, could it be leaner?

Maybe it might be better to construct manual types that satisfy the session types but have optional additional types?

### Test Strategy

The test strategy is:

1. Run `npx prisma migrate dev` using the schema the adaptor expects.  This makes sure the Prisma client has the types we expect. This creates a `test.db` file, which is git ignored.  That way running the tests won't introduce changes in PR's.
2. Create a Prisma Client
3. Run the battery of tests.

This seemed like the best approach, because you are testing an actual prisma client.  But I'd love feedback.

### Release candidate

For now I'm proposing this as a release candidate.  That way we can work with the adaptor in our templates to be sure it works nicely, and if any partners are keen to try this and provide feedback, they can knowing that it's only a release candidate.  Please let me know if I've done everything correctly.

I think we should probably work with this in our templates for a bit before doing a formal release?

## Type of change

- [ ] Patch: Bug (non-breaking change which fixes an issue)
- [ ] Minor: New feature (non-breaking change which adds functionality)
- [ ] Major: Breaking change (fix or feature that would cause existing functionality to not work as expected)

It's a major version, but it's brand new so not breaking.  So?

## Checklist

- [x] I have used `yarn changeset` to create a draft changelog entry (do NOT update the `CHANGELOG.md` files manually)
- [x] I have documented new APIs/updated the documentation for modified APIs (for public APIs)
